### ISSUE TITLE
HAML blocks and Rails 3

### ISSUE THREAD

**Issue by javierm** [jgdavey/tabletastic#6]
I'm using Rails 3.0 RC and tabletastic from master branch.

As you document, blocks in erb templates have changed for Rails 3. So the following code works:

```
<%= table_for @posts do |t|
  t.data :title
end %>
```

But the following code generates the table headers and body twice:

```
<%= table_for @posts do |t| %>
  <%= t.dada :title %>
<% end %>
```

I think it should work this way too, but no big deal anyway.

The problem comes when using HAML:

```
= table_for @posts do |t|
  = t.data :title
```

This generates the following output:

```
<thead><tr><th>Title/th></tr></thead>
<tbody>
<tr class="post odd" id="post_1"><td>My title</td></tr>
</tbody>
<table id="posts">0</table>
```

Using the block in one line works fine:
    = table_for(@posts) {|table| table.data(:title)}

But it's a bit awkward to use this sintax for more complex tables.

I've looked at HAML documentation and I think I'm using it right. I'm also using simple form this way:

```
= simple_form_for @post do |form|
  = form.input :title  
  = form.button :submit
```

And it works fine.

So I guess there's a bit of a problem with the way tabletastic handles blocks and the output buffer.

Thanks.

**Comment by jgdavey** [comment]
Thanks for this feedback!

You are correct, and I agree about the block concatenation. All three of the following forms should work in templates, without duplication:

```
# 1. One large concatted erb block
<%= table_for @posts do |t|
  t.data :title
end %>

# 2. Concatted block with settings in non-concatted erb expression
<%= table_for @posts do |t| %>
  <% t.data :title %>
<% end %>

# 3. Concatted block with settings in concatted erb expression
<%= table_for @posts do |t| %>
  <%= t.data :title %>
<% end %>
```

I believe that this commit should have straightened everything out: jgdavey/tabletastic@49a1fd0ff5b17b385e5e

Please check it out and let me know if this solves your issues as well. I'm not an active HAML user, so I'll take your word for it either way!

**Comment by javierm** [comment]
I'm afraid it still doesn't work, although the output has changed. Now it just generates:

```
<table id="posts">0</table>
```

Instead of generating the content of the table outside the tag, as it did before.

And I've got some more information. After experimenting for a while, I've found out that the following (notice the "-" instead of the "="):

```
= table_for(@posts) do |table|
  - table.data :title
```

Works fine. And it worked before the change, as well.

Somehow it reminds of how a CakePHP helper used to work: when doing "echo $session->flash()", it would print "1", whereas doing "$session->flash()" would print the actual error message. In that case it was because that method would return "true" of "false" and would use "echo" inside it.

I don't know much about the way buffers work in erb & haml, so it might be completely unrelated anyway.

**Comment by javierv** [comment]
After some playing around, based on an update on Railscasts #208 that says "capture (...) [works] outside of ERB views" , I've found out that everything seems to work changing the commit you mention slightly:

```
@@ -6,8 +6,7 @@ module Tabletastic
   klass = default_class_for(collection)
   options = args.extract_options!
   initialize_html_options(options, klass)
   result = ""
-      with_output_buffer { result = block.call(TableBuilder.new(collection, klass, self)) }
+      result = capture { block.call(TableBuilder.new(collection, klass, self)) }
   content_tag(:table, result, options[:html])
 end
```

I've also tried the following line:

```
result = with_output_buffer { block.call(TableBuilder.new(collection, klass, self)) }
```

It this case, HAML works as expected. However, as you might have also noticed, this breaks something. The following code doesn't generate the table contents:

```
<%= table_for(@posts) do |table|
  table.data(:actions => :all)
end %>
```

I currently have no clue about the way to test these behaviors, so all the testing I've done has been manual. I've also tried to look for documentation about the differences between "with_output_buffer" and "capture", but found nothing.

P.S: I had to change the account, but I'm the one who reported the issue.

**Comment by jgdavey** [comment]
<code>with_output_buffer</code> is actually not a public method in Rails, so I probably shouldn't have used it anyway. <code>capture</code>, on the other hand, is public, and (currently) utilizes <code>with_output_buffer</code> in its implemenation.

As it turns out, Rails 3 also deprecates block helpers that don't have the "=" sign. Since the <code>table.data</code> method accepts an optional block parameter, and seems to be outputting content, it really should called with an "=" sign, in both erb and haml, and not with a '-' in haml. Also this form would (and should) be valid:

  <%= table_for(@posts) do |table|
    table.data(:actions => :all)
  end %>

I'm accepting your patch above, and I confirm that it resolves this for all the (valid) scenarios. This is resolved in this commit: jgdavey/tabletastic@987a935f471aa

Thanks so much for your excellent bug report and help tracking it down, and your proposed fix.
